### ISSUE TITLE
Deploy Redash using Docker

### ISSUE THREAD

**Issue by odscjames** [open-contracting/deploy#1]
When we wrote these sripts, Redash provided a Ubuntu install script.

Now they don't - only a Docker version.

So this won't work on a fresh server. (A step fails with a 404 error)

There also raises worries that on the next major upgrade we might get into difficulties, as the upgrade script we are using won't be the right thing to do any more. (We use our own version, for reasons explained at the top of https://github.com/open-contracting/deploy/blob/master/salt/redash/upgrade-nointeraction )

**Comment by odscjames** [comment]
Redash V8 is coming soon, so we may have this problem shortly.

**Comment by jpmckinney** [comment]
Can we switch to the Docker version? It has its own [upgrade process](https://redash.io/help/open-source/admin-guide/how-to-upgrade) – not sure if we'd need to customize it similarly.

**Comment by odscjames** [comment]
I think it's wise to stick to the official install method as closely as possible, so I think it's looking like that's the first thing to try.

**Comment by odscjames** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Need to overhaul Redash Install process](https://trello.com/c/CvjzaiJg/111-need-to-overhaul-redash-install-process)

**Comment by jpmckinney** [comment]
I think we should consider the EC2 AMI as another option: https://redash.io/help/open-source/setup

**Comment by odscjames** [comment]
One drawback to using the EC2 AMI is we would lose a security feature we currently have; we currently put an Apache username and password in front of Redash, all users share the same one. This combined with the fact that we trust the limited number of users we have not to try and maliciously hack us means that there are some types of security hole that Redash might announce that we can treat as a normal upgrade and not urgent.

If we did consider AWS we should also do that as part of a wider review of hosting providers; while I personally know of no technical reason not to use AWS we should consider any ramifications for how we host, configure and support services.

**Comment by jpmckinney** [comment]
I don't think we need that extra layer of Basic Authentication. We don't do it for Redmine, which is far more mission critical.

The AWS AMI has the unique advantage of being a quickly deployable image maintained by the Redash developers. I don't think we need a wide review; it's just a choice between AWS or non-AWS (in which case we stay on Bytemark for simplicity).

**Comment by robredpath** [comment]
@jpmckinney as I understand it, the mode of operation for AWS is to have fundamentally disposable VMs (images from AMIs, which are updated by maintainers) and to keep persistent data in another storage medium (eg EBS, or one of their managed databases). Would we set up our Redash instance in that way, and keep it entirely separate from Salt? That seems like a good way to do it, just want to make sure we understand what's involved.

**Comment by odscjames** [comment]
@robredpath has had a look at the AMI, and it looks like it just sets up the Docker install process. Shall we just look at running Docker Redash on Bytemark?

I'm assuming the install process is nice and easy; the only thing that might be fiddly is making sure we take the database contents from the old install and put it into the new install (but that's true for any solution)

**Comment by jpmckinney** [comment]
Thanks for looking into those details. If the AMI is just adding to Docker Redash, then I agree that we might as well keep things simple and use Docker Redash directly on Bytemark.

**Comment by jpmckinney** [comment]
I removed content from `ocds-redash.sls` that no longer works due to the lack of `bootstrap.sh`. Then I ran:

```
salt-ssh 'ocds-redash' file.remove /opt/redash/current/bin/upgrade-nointeraction
salt-ssh 'ocds-redash' pkg.purge python-requests,python-semver
```

**Comment by jpmckinney** [comment]
I also removed Basic Authentication, then ran:

```
salt-ssh 'ocds-redash' file.remove /etc/nginx/redash-htpasswd
```

**Comment by odscjames** [comment]
Redash V8 is now out.

**Comment by odscjames** [comment]
STATUS UPDATE:

Have created new server, live2.redash - redash here is installed via Docker.

I checked the salt  into master branch so that master has the new server, and it reflects current state.

As you can see in https://github.com/open-contracting/deploy/blob/master/salt/ocds-redash.sls there are some manual steps documented in comments - we need to decide later how much effort to put into this.

I tried a process of migrating postgres database contents to new server. 

I had to give my self admin rights by editing the database contents by hand - sysadmins at ODSC will need admin access to do various work and checks.

First problem: Something is wrong with data sources and it won't let me work with them. I need to work on this.

Also todo: add to Prometheus

**Comment by jpmckinney** [comment]
I think it's fine for the deployment to be manual, since we don't need to upgrade more than, say, once a year. In that case, it'd be better to have the commented steps as a docs page.

If any of your comments above aren't already in the Salt file, those should be added to the eventual docs page as well.

**Comment by odscjames** [comment]
Problem sorted. Added to Prometheus. Also did an upgrade from 7 to 8 on new server so know that works.

http://live2.redash.opencontracting.uk0.bigv.io:9090 should now be live with same users and data

TODO
* Can you make my account an admin on the new server? This will help save time on any future work.
* Ticket in CRM to tell all staff; do we need to alert external users? Looking at the users page most of them have not logged in for months.
* Move Salt comments to a docs page
* Turn off old server in a few weeks

Does that sound good @jpmckinney ?

**Comment by odscjames** [comment]
Just got low memory alert; I put a silence on 20% alarm for that server only. We'll monitor what it's normal memory use is.

**Comment by jpmckinney** [comment]
Sounds good! Comments below.

> Can you make my account an admin on the new server? This will help save time on any future work.

Done.

> Ticket in CRM to tell all staff; do we need to alert external users? Looking at the users page most of them have not logged in for months.

I added a CNAME from redash.kingfisher.open-contracting.org to live2.redash.opencontracting.uk0.bigv.io

Can we serve Redash from that domain without the port number? That way, we won't have to alert everyone every time we upgrade the server.

I added an issue to enable HTTPS: #124 

Please also ask helpdesk analysts to contact people at the discretion.

> Move Salt comments to a docs page

Let's include a link to this page: https://redash.io/help/open-source/admin-guide/how-to-upgrade

----

Separately: #78 is now unblocked. I also added #123 as a follow-up issue.

**Comment by jpmckinney** [comment]
> Just got low memory alert; I put a silence on 20% alarm for that server only. We'll monitor what it's normal memory use is.

Should we just scale to 3GB now? The old server was already frequently at 70%: https://monitor.prometheus.open-contracting.org/graph?g0.range_input=1d&g0.stacked=0&g0.expr=1%20-%20node_memory_MemAvailable_bytes%20%2F%20node_memory_MemTotal_bytes&g0.tab=0

**Comment by odscjames** [comment]
I will remove Prometheus from that server and upgrade to 3G now. (Moving the app to port 80 means we can't use the Apache Prometheus client as it currently stands)

**Comment by odscjames** [comment]
Done and the app is now on port 80. I'll work out how to put Prometheus back next.

**Comment by odscjames** [comment]
Sorry for friday afternoon faff. Came across something to sort out, and it turned out getting it running on port 80 was more difficult than I expected.

The app is now available via http://redash.open-contracting.org and the server is back in Promethous.

Apache proxies traffic from the domain name to the actual app. This means we can add SSL at the Apache layer using our existing salt scripts, and I'm hoping that will be an easy operation.

TODO:
* Ticket in CRM to tell all staff
* Move Salt comments to a docs page
* Turn off old server in a few weeks

**Comment by jpmckinney** [comment]
Sounds good!

I know the `DocumentRoot` in `redash.conf` doesn't matter, but it's confusing that it refers to `ocdskingfisherprocess`, so we should either remove the directive or point it to some dummy location.

To check my understanding:

- Apache proxies redash.open-contracting:80 to Nginx and prom-client.…:80 to Prometheus.
- The Docker image runs Nginx to serve requests to Redash.
- Nginx is not exposed to the world.

Is that correct?

**Comment by odscjames** [comment]
Yes, except Redash is exposed to the world as Docker is exposing ports to 0.0.0.0. Think we can docker port map to 127.0.0.1 but also unless Apache was adding any extra security (and it's not currently) there is no difference anyway. We are just using Apache to provide the virtual host functionality for 2 sites and a SSL end point.

(Actually looking at it, it might be the case that the Nginx Docker is just proxing anyway, in which case we could remove it and have Apache proxy direct to the Redash server. But it would take more time to work out, and unless it causes a problem ......)

**Comment by jpmckinney** [comment]
Ah, yes, I can access redash.open-contracting.org:9090

I don't think it matters that we have some extra traffic on the loopback device, so having Apache proxy to Nginx instead of to Redash is fine.

**Comment by jpmckinney** [comment]
Once we turn off the old server, I think we can:

* [ ] Delete salt/nginx/* and salt/prometheus-client-nginx
* [ ] Remove the reference from docs/deploy/prometheus.rst
* [ ] Remove client_password_as_nginx_file from Pillar

**Comment by odscjames** [comment]
I think this is now at a state where I can alert analysts and then we'll leave it for a week and make sure they are ok. Will make CRM ticket now.

Also add to TODO list:
* Add URL of loginpage to uptime robot?

**Comment by jpmckinney** [comment]
Added new server and removed old server from Uptime Robot.

**Comment by odscjames** [comment]
Removed old server from Prom and stopped it. If no-one complains, in a bit will delete old server and do final cleanup.

**Comment by odscjames** [comment]
@jpmckinney  I did these docs - do they look good?

I did intend to use a pull request - not using one was a mistake on my part, sorry.

**Comment by odscjames** [comment]
Server deleted.

**Comment by odscjames** [comment]
The 2 pull requests above should take care of all the todo's and lose ends, then this can be closed.

**Comment by jpmckinney** [comment]
I'll review the docs commit when I have time. Otherwise, all done!
